libraries/chain/finality_core.cpp

```diff
@@ -78,6 +78,29 @@ qc_claim_t finality_core::latest_qc_claim() const
    return qc_claim_t{.block_num = links.back().target_block_num, .is_strong_qc = links.back().is_link_strong};
 }
 
+/**
+ *  @pre  all finality_core invariants
+ *  @post same
+ *  @returns timestamp of latest qc_claim made by the core
+ */
+block_time_type finality_core::latest_qc_block_timestamp() const {
+   return get_block_reference(links.back().target_block_num).timestamp;
+}
+
+/**
+ *  @pre  all finality_core invariants
+ *  @post same
+ *  @returns boolean indicating whether `id` is an ancestor of this block
+ */
+bool finality_core::extends(const block_id_type& id) const {
+   uint32_t block_num = block_header::num_from_id(id);
+   if (block_num >= last_final_block_num() && block_num < current_block_num()) {
```

Comment:
When `block_num < last_final_block_num()`, it can be an ancestor. But it is too old and out of the links. Maybe add a comment.

Reply:
It doesn't matter that it is out of the link, it is a safety feature of the finalizer that we can only vote on proposals which are descendants of the one we are locked on.